Re-integrate next_infra into develop

joshmoore opened the pull request:
# What this PR does

Re-integrates the `next_infra` branch back into `develop`. Once merged, the `next_infra` integration branch will be deleted. The following PRs will then be included for release in 5.4.2:

## Users
### CLI
 * #5607: `bin/omero import --target` fix
 * #5610: `bin/omero import --debug` fix
 * #5608: `bin/omero admin mail` timeout improvements `(*)`
 * #5538: `bin/omero admin log` added command `(*)`
## Devs
### Python
 * #5595: new omero-marshal (0.5.2) with masks support
 * #5585: Ship `omero_setup.py` for downstream build systems `(*)`
 * #5573: Refactor test classes (scripts) for downstream build systems
 * #5548: Refactor test classes (CLI) for downstream build systems
 * Note: the build system changes were designed to permit integration tests in downstream projects, see:
    - https://github.com/openmicroscopy/omero-test-infra
    - https://github.com/ome/omero-mapr/pull/23
    - https://github.com/ome/omero-cli-render/pull/1
    - https://github.com/ome/rOMERO-gateway/pull/22
    - https://github.com/ome/scripts/pull/136
### Java
 * #5490: add methods for updating OMERO.tables `(*)`
 * #5513: fixes for the Java Gateway (sessions) `(*)`
 * #5545: fixes for the Java Gateway (rendering)

----

`(*): includes new API that should be signed off on`

joshmoore committed:
render/gateway: close as many rendering engines as possible

joshmoore committed:
gateway: track created services

A new "tracked service" internal API has been added. All calls to
`_create_func` in `ProxyObjectWrapper` now registers the string
representation of the created service as well as the current call
stack to the owning `BlitzGateway` instance. If on deletion, services
are left open, then an ERROR message including the call stack will
be printed.

joshmoore committed:
Fix flake8

dominikl committed:
Added update method

dominikl committed:
javadoc: add note that one can't change table size

dominikl committed:
Remove usage of guest session

dominikl committed:
Also throw DSOutOfServiceException on connection loss

dominikl committed:
Add join session 'LoginCredentials' and test

dominikl committed:
Fix: Don't swallow Exception for Gateway.getSessionId()

dominikl committed:
Add accidentaly commented out tests again

dominikl committed:
Use Java's UUID method instead of regex

dominikl committed:
Catch exception

dominikl committed:
Make Gateway AutoClosable

mtbc committed:
add a "log" command to RawAccessRequest

dominikl committed:
Remove stale connectors

dominikl committed:
Check arguments

mtbc committed:
differently name manual RawAccessRequest logger

Avoids potential for confusion with messages genuinely
from the RawAccessRequest class' logger.

dominikl committed:
getHierarchy without ids is a valid request (~ 'get all')

dominikl committed:
Just clarified doc for getHierarchy method

dominikl committed:
Removed printStackTrace from integration test

dominikl committed:
Adjusted javadoc comment

jburel committed:
Add new method

jburel committed:
Add documentation

jburel committed:
Add tests

jburel committed:
flake8

jburel committed:
Add tests for register/unregister method

jburel committed:
add test for _closeRE method

jburel committed:
Add doc

jburel committed:
Make sure rendering engines are closed

jburel committed:
Introduce new file in testlib and deprecate one

jburel committed:
Update tests

jburel committed:
Move the script to testlib

jburel committed:
Move the tests to the ome/scripts repo

dominikl committed:
Add test for admin cli 'log' command

dominikl committed:
Add the 'log' functionality to the admin cli

jburel committed:
Merge remote-tracking branch 'origin/develop' into test-scripts

mtbc committed:
Merge pull request #16 from dominikl/admin_create_log

Add the "log" command to admin CLI, discussed with @joshmoore

mtbc committed:
flake8 fix

joshmoore committed:
Merge pull request #5545 from jburel/gateway-methods

Gateway methods

joshmoore committed:
Merge pull request #5513 from dominikl/java_gateway_fixes

Java gateway bugfixes

joshmoore committed:
Merge pull request #5548 from jburel/test-cli

Test cli

joshmoore committed:
Merge pull request #5573 from jburel/test-scripts

Test scripts

joshmoore committed:
Merge pull request #5490 from dominikl/add_update_tables

Add update tables method

joshmoore committed:
Merge branch 'clisplit' into log-cmd

joshmoore committed:
Merge branch 'clisplit' into log-cmd

joshmoore committed:
Rename tools/test_setup.py to OmeroPy/src/omero_setup.py

So that downstream repositories can make use of the PyTest
class defined here, the module will need to be packaged as
part of lib/python.

joshmoore committed:
omero_setup: Add src to end of PYTHONPATH

By adding src to the classpath at the beginning,
modules that were not present under src/, having
been added later under target/, were being shadowed.

joshmoore committed:
Merge pull request #5585 from joshmoore/test_omero_setup

Rename tools/test_setup.py to OmeroPy/src/omero_setup.py

mtbc committed:
move omero.grid.LogLevel to omero.constants.LogLevel

define in Constants.ice rather than Repositories.ice

mtbc committed:
deprecate RawAccessRequest

Intended only for internal use, testing and diagnostics.
Good candidate for early microservice.

joshmoore committed:
Merge pull request #5538 from mtbc/log-cmd

add a "log" command to RawAccessRequest

sbesson committed:
Bump omero-marshal to 0.5.2

This release adds support for Masks - see https://github.com/openmicroscopy/omero-marshal/blob/v0.5.2/CHANGELOG.md

sbesson committed:
Add Mask shape to the fixture of the JSON API ROI integration tests

mtbc committed:
prevent import targeting a container where canLink is false

mtbc committed:
extend timeout for sending e-mail from CLI

mtbc committed:
allow CLI admin email to have --wait argument set

mtbc committed:
apply selected log level to SciJava native library loader

mtbc committed:
tidy class header

joshmoore committed:
Merge pull request #5608 from mtbc/extend-email-timeout

extend CLI admin email timeout

joshmoore committed:
Merge pull request #5607 from mtbc/import-target

have import fail fast if into illegal container

joshmoore committed:
Merge pull request #5610 from mtbc/fix-loglevel

allow SciJava native lib loader loglevel to be configured

joshmoore committed:
Merge pull request #5595 from sbesson/omero_marshal_0.5.2

Bump omero-marshal to 0.5.2

joshmoore commented:
History of the PRs included here:

```
0fabfb0e36 (origin/next_infra) Merge pull request #5595 from sbesson/omero_marshal_0.5.2 (19 hours ago) <Josh Moore>
2b7b7fe03e Merge pull request #5610 from mtbc/fix-loglevel (25 hours ago) <Josh Moore>
1c9a8d97d1 Merge pull request #5607 from mtbc/import-target (25 hours ago) <Josh Moore>
622b27da71 Merge pull request #5608 from mtbc/extend-email-timeout (4 days ago) <Josh Moore>
077384b7e4 Merge pull request #5538 from mtbc/log-cmd (6 weeks ago) <Josh Moore>
8b8be4ea7d (origin/clisplit) Merge pull request #5585 from joshmoore/test_omero_setup (7 weeks ago) <Josh Moore>
3a9dedaef1 Merge pull request #5490 from dominikl/add_update_tables (8 weeks ago) <Josh Moore>
e4a1a63afa Merge pull request #5573 from jburel/test-scripts (8 weeks ago) <Josh Moore>
54d6f3bdb3 Merge pull request #5548 from jburel/test-cli (8 weeks ago) <Josh Moore>
48d5860ebc Merge pull request #5513 from dominikl/java_gateway_fixes (8 weeks ago) <Josh Moore>
b476a51f22 Merge pull request #5545 from jburel/gateway-methods (8 weeks ago) <Josh Moore>
```

jburel commented on components/blitz/src/omero/gateway/Gateway.java:
```diff
@@ -1220,7 +1234,9 @@ private void keepSessionAlive() throws DSOutOfServiceException {
             if (c.needsKeepAlive()) {
                 if (!c.keepSessionAlive()) {
-                    throw new DSOutOfServiceException("Network not available");
+                    // Session has died, e. g. due to server restart.
```
@dominikl how does insight handle the fact that no error is thrown now?

dominikl replied:
Should work as before (show connection warning and reconnect again, when possible), tested on the original PR #5513 (point 3). 

As far as I remember the Exception was the problem, because this reconnect mechanism only worked for a stale network connection, but not when the server was restarted. In that case even when the server connection is back, keepSessionAlive fails, and the exception crashed Insight. With this change the stale connector is quietly removed and a new connector will be created.

jburel commented on components/blitz/src/omero/gateway/facility/AdminFacility.java:
```diff
@@ -392,6 +404,9 @@ public void setAdminPrivileges(SecurityContext ctx, ExperimenterData user,
     public void addAdminPrivileges(SecurityContext ctx, ExperimenterData user,
             Collection<String> privileges) throws DSOutOfServiceException,
             DSAccessException {
+        if(!Pojos.hasID(user) || CollectionUtils.isEmpty(privileges))
```
space between ``if (``

jburel commented on components/blitz/src/omero/gateway/facility/AdminFacility.java:
```diff
@@ -421,6 +436,8 @@ public void addAdminPrivileges(SecurityContext ctx, ExperimenterData user,
     public void removeAdminPrivileges(SecurityContext ctx,
             ExperimenterData user, Collection<String> privileges)
             throws DSOutOfServiceException, DSAccessException {
+        if(!Pojos.hasID(user) || CollectionUtils.isEmpty(privileges))
```
same

jburel commented on components/blitz/src/omero/gateway/facility/BrowseFacility.java:
```diff
@@ -108,23 +113,32 @@
     /**
-     * Retrieves hierarchy trees rooted by a given node.
-     * i.e. the requested node as root and all of its descendants.
+     * Retrieves hierarchy trees rooted by a given node. i.e. the requested node
```
extra ``.`` before ``i.e.``

jburel commented on components/blitz/src/omero/gateway/facility/MetadataFacility.java:
```diff
@@ -83,6 +87,9 @@
      */
     public ImageAcquisitionData getImageAcquisitionData(SecurityContext ctx,
             long imageId) throws DSOutOfServiceException, DSAccessException {
+        if(imageId < 0)
```
space after ``if``

jburel commented on components/blitz/src/omero/gateway/facility/MetadataFacility.java:
```diff
@@ -106,7 +113,9 @@ public ImageAcquisitionData getImageAcquisitionData(SecurityContext ctx,
             throws DSOutOfServiceException, DSAccessException {
         List<ChannelData> result = new ArrayList<ChannelData>();
-
+        if(imageId < 0)
```
same

jburel commented on components/blitz/src/omero/gateway/facility/ROIFacility.java:
```diff
@@ -941,6 +960,9 @@ else if (shape.isDirty()) {
     public Collection<ROIResult> loadROIsForFolder(SecurityContext ctx,
             long imageId, long folderId) throws DSOutOfServiceException,
             DSAccessException {
+        if (imageId < 0 && folderId < 0)
```
Shouldn't be or?

jburel commented on components/blitz/src/omero/gateway/facility/TablesFacility.java:
```diff
@@ -263,6 +268,9 @@ public TableData getTable(SecurityContext ctx, long fileId, long rowFrom,
     public long[] query(SecurityContext ctx, long fileId, String condition,
             long start, long stop, long step) throws DSOutOfServiceException,
             DSAccessException {
+        if (fileId < 0)
```
Other methods return ``null`` when the id is ``<0``
why is it different for that method?

dominikl replied:
That's true, returning `null` would be better. I generally wasn't very consistent, sometimes tried to not introduce potential NullPointerExceptions, but then in other cases return null or throw IllegalArgumentExceptions. Could that be unified throughout the gateway at some point (could easily break clients though)?

jburel commented on components/blitz/src/omero/gateway/facility/TablesFacility.java:
```diff
@@ -521,5 +538,62 @@ public TableData getTable(SecurityContext ctx, long fileId, long rowFrom,
+            throws DSOutOfServiceException, DSAccessException {
+        TablePrx table = null;
+        try {
+            if (data.getOriginalFileId() == -1)
```
``<0`` instead?

jburel commented on components/blitz/src/omero/gateway/facility/TablesFacility.java:
```diff
@@ -521,5 +538,62 @@ public TableData getTable(SecurityContext ctx, long fileId, long rowFrom,
+
+            table = sr.openTable(file);
+
+            long[] colIndex = new long[data.getColumns().length];
```
maybe add a variable ``cols = data.getColumns`` to reduce the number of calls

jburel commented on components/blitz/src/omero/gateway/facility/TablesFacilityHelper.java:
```diff
@@ -225,6 +226,8 @@ void parseData(Data data, TableDataColumn[] header) {
                 for (int j = 0; j < nRows; j++) {
                     MaskData md = new MaskData(mc.x[j], mc.y[j], mc.w[j],
                             mc.h[j], mc.bytes[j]);
+                    md.setZ(mc.theZ[j]);
```
Will ``theZ`` and ``theT`` always be filled?

dominikl replied:
I'd expect that both are always arrays of `nRows` size, but you're right, I guess the values could be `-1` if `z` and `t` are not explicitly set. Which is actually a bug in `ShapeData` (`getZ` and `getT` should return `0` not `-1` in that case).  Follow-up PR maybe?

dominikl replied:
Or does `-1` for z/t mean "all" z/t ? As `channel` can also be `-1`. 
The getters and setters in `ShapeData` are all inconsistent, e. g. if `z` is not set, `getZ()` will return `-1`, but if you `setZ(-1)` actually `0` is set!

I think I leave it as it is for now.

jburel replied:
don't modify in this PR
Changing the getter and setter should be done in another PR

jburel commented on components/blitz/src/omero/gateway/facility/TablesFacilityHelper.java:
```diff
@@ -467,4 +470,198 @@ int getNRows() {
+        if (toUpdate.rowNumbers.length != data.getData()[0].length)
+            throw new IllegalArgumentException("Row size is different!");
+
+        for (int c = 0; c < data.getColumns().length; c++) {
```
same ``cols = data.getColumns()`` 
and for ``data.getData()``

jburel commented on components/tools/OmeroJava/test/integration/gateway/TablesFacilityTest.java:
```diff
@@ -276,5 +281,46 @@ public void testThreshold() throws Exception {
     }
+    
+    @Test(dependsOnMethods = { "testThreshold" })
+    public void testUpdateTable() throws Exception {
```
``cols= td.getColumns()`` same for data 
since the various will have an effect on performance

joshmoore replied:
I'll leave @dominikl to respond, but I think these are all local invocations since it's `omero.gateway.model.TableData` and not `omero.api.TablesPrx`.

jburel replied:
in that case no need to make changes
I got lost in the PR review

dominikl replied:
Yes, just accesses the `columns` array in the `TableData` object.

jburel commented on components/tools/OmeroPy/src/omero/gateway/__init__.py:
```diff
@@ -1568,6 +1569,51 @@ def __init__(self, username=None, passwd=None, client_obj=None, group=None,
+
+    def _unregister_service(self, service_string):
+        """
+        Called on close of a service.
```
``Call``?

jburel commented on components/tools/OmeroPy/test/integration/clitest/cli.py:
```diff
@@ -94,6 +104,9 @@ class ArgumentFixture(object):
     """
     Used to test the user/group argument
     """
+    warnings.warn("Deprecated in 5.4.1."
```
message should be 5.4.2 now

joshmoore committed:
End trailing whitespace

joshmoore committed:
Fix suggestions from jburel

joshmoore committed:
Merge pull request #5618 from joshmoore/next_infra_fixes

Next infra fixes

joshmoore committed:
Correct more 'deprecated in 5.4.1' lines

joshmoore committed:
scripts: Exclude setup.py and test/**

Without the exclusion, python files appear under
the OMERO scripts menu which lead to exceptions
("No params found").

joshmoore commented:
Based on today's greenedness, merging to prepare for release. This is a last call for objections to the API additions `(*)` noted in the description.
